In [1]:
import pandas as pd
import numpy as np


In [2]:
data= pd.read_csv(r"C:\Users\SNEHAL\Desktop\data anlyst file\file.csv")
data

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa               Weather  
0               8.0     101.24                   Fog  
1               8.0     101.24                   Fog  
2               4.0     101.26  Freezing Drizzle,Fog  
3               4.0     101.27  Freezing Drizzle,Fog  
4               4.8     101.23                   Fog  
...             ...        ...                   ...  
8779            9.7     100.13                  Snow  
8780            9.7     100.03                  Snow  
8781            4.8      99.95                  Snow  
8782            9.7      99.91                  Snow  
8783           11.3      99.89                  Snow  

[8784 rows x 8 columns]

In [3]:
data.shape

(8784, 8)

In [4]:
data.index

RangeIndex(start=0, stop=8784, step=1)

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8784 entries, 0 to 8783
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date/Time         8784 non-null   object 
 1   Temp_C            8784 non-null   float64
 2   Dew Point Temp_C  8784 non-null   float64
 3   Rel Hum_%         8784 non-null   int64  
 4   Wind Speed_km/h   8784 non-null   int64  
 5   Visibility_km     8784 non-null   float64
 6   Press_kPa         8784 non-null   float64
 7   Weather           8784 non-null   object 
dtypes: float64(4), int64(2), object(2)
memory usage: 549.1+ KB


In [6]:
data.describe()

Temp_C  Dew Point Temp_C    Rel Hum_%  Wind Speed_km/h  \
count  8784.000000       8784.000000  8784.000000      8784.000000   
mean      8.798144          2.555294    67.431694        14.945469   
std      11.687883         10.883072    16.918881         8.688696   
min     -23.300000        -28.500000    18.000000         0.000000   
25%       0.100000         -5.900000    56.000000         9.000000   
50%       9.300000          3.300000    68.000000        13.000000   
75%      18.800000         11.800000    81.000000        20.000000   
max      33.000000         24.400000   100.000000        83.000000   

       Visibility_km    Press_kPa  
count    8784.000000  8784.000000  
mean       27.664447   101.051623  
std        12.622688     0.844005  
min         0.200000    97.520000  
25%        24.100000   100.560000  
50%        25.000000   101.070000  
75%        25.000000   101.590000  
max        48.300000   103.650000

In [7]:
data.dtypes

Date/Time            object
Temp_C              float64
Dew Point Temp_C    float64
Rel Hum_%             int64
Wind Speed_km/h       int64
Visibility_km       float64
Press_kPa           float64
Weather              object
dtype: object

In [8]:
data['Weather'].unique()

array(['Fog', 'Freezing Drizzle,Fog', 'Mostly Cloudy', 'Cloudy', 'Rain',
       'Rain Showers', 'Mainly Clear', 'Snow Showers', 'Snow', 'Clear',
       'Freezing Rain,Fog', 'Freezing Rain', 'Freezing Drizzle',
       'Rain,Snow', 'Moderate Snow', 'Freezing Drizzle,Snow',
       'Freezing Rain,Snow Grains', 'Snow,Blowing Snow', 'Freezing Fog',
       'Haze', 'Rain,Fog', 'Drizzle,Fog', 'Drizzle',
       'Freezing Drizzle,Haze', 'Freezing Rain,Haze', 'Snow,Haze',
       'Snow,Fog', 'Snow,Ice Pellets', 'Rain,Haze', 'Thunderstorms,Rain',
       'Thunderstorms,Rain Showers', 'Thunderstorms,Heavy Rain Showers',
       'Thunderstorms,Rain Showers,Fog', 'Thunderstorms',
       'Thunderstorms,Rain,Fog',
       'Thunderstorms,Moderate Rain Showers,Fog', 'Rain Showers,Fog',
       'Rain Showers,Snow Showers', 'Snow Pellets', 'Rain,Snow,Fog',
       'Moderate Rain,Fog', 'Freezing Rain,Ice Pellets,Fog',
       'Drizzle,Ice Pellets,Fog', 'Drizzle,Snow', 'Rain,Ice Pellets',
       'Drizzle,Snow,Fog', 

In [9]:
#.nunique() It shows total no of unique values in each column.it can be applied on a single column as well as on whole dataframe
data.nunique()

Date/Time           8784
Temp_C               533
Dew Point Temp_C     489
Rel Hum_%             83
Wind Speed_km/h       34
Visibility_km         24
Press_kPa            518
Weather               50
dtype: int64

In [10]:
#.count()   It shows the total no of  non-null values in each column. it can be applied on a single column as well as on whole dataframe
data.count()

Date/Time           8784
Temp_C              8784
Dew Point Temp_C    8784
Rel Hum_%           8784
Wind Speed_km/h     8784
Visibility_km       8784
Press_kPa           8784
Weather             8784
dtype: int64

In [11]:
# . value_counts    in a column, it shows all the unique values with their count. it can be applied on single column only.
data['Weather'].value_counts()

Weather
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                

# Q.1 find all unique 'wind speed' values in data?


In [12]:
data['Wind Speed_km/h'].unique()

array([ 4,  7,  6,  9, 15, 13, 20, 22, 19, 24, 30, 35, 39, 32, 33, 26, 44,
       43, 48, 37, 28, 17, 11,  0, 83, 70, 57, 46, 41, 52, 50, 63, 54,  2],
      dtype=int64)

# Q.2 Find the no of times when the 'weather is exactly clear'?

In [13]:
data['Weather'].value_counts()

Weather
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                

In [14]:
#filtering
fdata=data[data['Weather'] =='Clear']
fdata

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67      1/3/2012 19:00   -16.9             -24.8         50               24   
114     1/5/2012 18:00    -7.1             -14.4         56               11   
115     1/5/2012 19:00    -9.2             -15.4         61                7   
116     1/5/2012 20:00    -9.8             -15.7         62                9   
117     1/5/2012 21:00    -9.0             -14.8         63               13   
...                ...     ...               ...        ...              ...   
8646   12/26/2012 6:00   -13.4             -14.8         89                4   
8698  12/28/2012 10:00    -6.1              -8.6         82               19   
8713   12/29/2012 1:00   -11.9             -13.6         87               11   
8714   12/29/2012 2:00   -11.8             -13.1         90               13   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather  
67             25.0     101.74   Clear  
114            25.0     100.71   Clear  
115            25.0     100.80   Clear  
116            25.0     100.83   Clear  
117            25.0     100.83   Clear  
...             ...        ...     ...  
8646           25.0     102.47   Clear  
8698           24.1     101.27   Clear  
8713           25.0     101.31   Clear  
8714           25.0     101.33   Clear  
8756           25.0     101.52   Clear  

[1326 rows x 8 columns]

In [15]:
gb= data.groupby('Weather').get_group('Clear')
gb

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67      1/3/2012 19:00   -16.9             -24.8         50               24   
114     1/5/2012 18:00    -7.1             -14.4         56               11   
115     1/5/2012 19:00    -9.2             -15.4         61                7   
116     1/5/2012 20:00    -9.8             -15.7         62                9   
117     1/5/2012 21:00    -9.0             -14.8         63               13   
...                ...     ...               ...        ...              ...   
8646   12/26/2012 6:00   -13.4             -14.8         89                4   
8698  12/28/2012 10:00    -6.1              -8.6         82               19   
8713   12/29/2012 1:00   -11.9             -13.6         87               11   
8714   12/29/2012 2:00   -11.8             -13.1         90               13   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather  
67             25.0     101.74   Clear  
114            25.0     100.71   Clear  
115            25.0     100.80   Clear  
116            25.0     100.83   Clear  
117            25.0     100.83   Clear  
...             ...        ...     ...  
8646           25.0     102.47   Clear  
8698           24.1     101.27   Clear  
8713           25.0     101.31   Clear  
8714           25.0     101.33   Clear  
8756           25.0     101.52   Clear  

[1326 rows x 8 columns]

# Q.3 Find the no of times when the wind speed was exactly 4km/hr?

In [16]:
f_data= data[data['Wind Speed_km/h']== 4]
f_data


Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
96       1/5/2012 0:00    -8.8             -11.7         79                4   
101      1/5/2012 5:00    -7.0              -9.5         82                4   
146      1/7/2012 2:00    -8.1             -11.1         79                4   
...                ...     ...               ...        ...              ...   
8768   12/31/2012 8:00    -8.6             -10.3         87                4   
8769   12/31/2012 9:00    -8.1              -9.6         89                4   
8770  12/31/2012 10:00    -7.4              -8.9         89                4   
8772  12/31/2012 12:00    -5.8              -7.5         88                4   
8773  12/31/2012 13:00    -4.6              -6.6         86                4   

      Visibility_km  Press_kPa       Weather  
0               8.0     101.24           Fog  
1               8.0     101.24           Fog  
96              9.7     100.32          Snow  
101             4.0     100.19          Snow  
146            19.3     100.15        Cloudy  
...             ...        ...           ...  
8768            3.2     101.14  Snow Showers  
8769            2.4     101.09          Snow  
8770            6.4     101.05      Snow,Fog  
8772           12.9     100.78          Snow  
8773           12.9     100.63          Snow  

[474 rows x 8 columns]

In [17]:
gf= data.groupby('Wind Speed_km/h').get_group(4)
gf

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
96       1/5/2012 0:00    -8.8             -11.7         79                4   
101      1/5/2012 5:00    -7.0              -9.5         82                4   
146      1/7/2012 2:00    -8.1             -11.1         79                4   
...                ...     ...               ...        ...              ...   
8768   12/31/2012 8:00    -8.6             -10.3         87                4   
8769   12/31/2012 9:00    -8.1              -9.6         89                4   
8770  12/31/2012 10:00    -7.4              -8.9         89                4   
8772  12/31/2012 12:00    -5.8              -7.5         88                4   
8773  12/31/2012 13:00    -4.6              -6.6         86                4   

      Visibility_km  Press_kPa       Weather  
0               8.0     101.24           Fog  
1               8.0     101.24           Fog  
96              9.7     100.32          Snow  
101             4.0     100.19          Snow  
146            19.3     100.15        Cloudy  
...             ...        ...           ...  
8768            3.2     101.14  Snow Showers  
8769            2.4     101.09          Snow  
8770            6.4     101.05      Snow,Fog  
8772           12.9     100.78          Snow  
8773           12.9     100.63          Snow  

[474 rows x 8 columns]

# answer= The wind speed was exactly 4 km/hr 474 times

# Q .4 Find out all the null value in data

In [18]:
data.isnull().sum()

Date/Time           0
Temp_C              0
Dew Point Temp_C    0
Rel Hum_%           0
Wind Speed_km/h     0
Visibility_km       0
Press_kPa           0
Weather             0
dtype: int64

# Q.5 rename the column name"weather" to 'weather conditions"

In [19]:
data = data.rename(columns={'Weather': 'Weather Condition'})
data

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa     Weather Condition  
0               8.0     101.24                   Fog  
1               8.0     101.24                   Fog  
2               4.0     101.26  Freezing Drizzle,Fog  
3               4.0     101.27  Freezing Drizzle,Fog  
4               4.8     101.23                   Fog  
...             ...        ...                   ...  
8779            9.7     100.13                  Snow  
8780            9.7     100.03                  Snow  
8781            4.8      99.95                  Snow  
8782            9.7      99.91                  Snow  
8783           11.3      99.89                  Snow  

[8784 rows x 8 columns]

# Q.6 what is the mean visibility?  answer- 27.66 is mean

In [20]:
mean= data['Visibility_km'].mean()
mean

27.664446721311478

# Q.7 what is the standard deviation of pressure  in this data?   answer- 0.84

In [21]:
std= data['Press_kPa'].std()
std

0.8440047459486483

# Q.8 WHat is the variance in 'relative humidity' in this data Answer= 286.24


In [22]:
data['Rel Hum_%'].var()

286.24855019850196

# Q.9 Find all instance when 'snow' was recorded? ans- 583 times

In [23]:
data.columns

Index(['Date/Time', 'Temp_C', 'Dew Point Temp_C', 'Rel Hum_%',
       'Wind Speed_km/h', 'Visibility_km', 'Press_kPa', 'Weather Condition'],
      dtype='object')

In [24]:
data['Weather Condition'].value_counts()

Weather Condition
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle      

In [25]:
data1= data[data['Weather Condition']=='Snow']
data1

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
55       1/3/2012 7:00   -14.0             -19.5         63               19   
84      1/4/2012 12:00   -13.7             -21.7         51               11   
86      1/4/2012 14:00   -11.3             -19.0         53                7   
87      1/4/2012 15:00   -10.2             -16.3         61               11   
88      1/4/2012 16:00    -9.4             -15.5         61               13   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa Weather Condition  
55             25.0     100.95              Snow  
84             24.1     101.25              Snow  
86             19.3     100.97              Snow  
87              9.7     100.89              Snow  
88             19.3     100.79              Snow  
...             ...        ...               ...  
8779            9.7     100.13              Snow  
8780            9.7     100.03              Snow  
8781            4.8      99.95              Snow  
8782            9.7      99.91              Snow  
8783           11.3      99.89              Snow  

[390 rows x 8 columns]

In [26]:
# str.contains snow
data1=data[data['Weather Condition'].str.contains('Snow')]
data1

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
41      1/2/2012 17:00    -2.1              -9.5         57               22   
44      1/2/2012 20:00    -5.6             -13.4         54               24   
45      1/2/2012 21:00    -5.8             -12.8         58               26   
47      1/2/2012 23:00    -7.4             -14.1         59               17   
48       1/3/2012 0:00    -9.0             -16.0         57               28   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa Weather Condition  
41             25.0      99.66      Snow Showers  
44             25.0     100.07      Snow Showers  
45             25.0     100.15      Snow Showers  
47             19.3     100.27      Snow Showers  
48             25.0     100.35      Snow Showers  
...             ...        ...               ...  
8779            9.7     100.13              Snow  
8780            9.7     100.03              Snow  
8781            4.8      99.95              Snow  
8782            9.7      99.91              Snow  
8783           11.3      99.89              Snow  

[583 rows x 8 columns]

# Q.10 Find all instances when 'wind speed' is above 24 and visiblity 25

In [27]:
df1=data[(data['Wind Speed_km/h']>=24) & (data['Visibility_km']==25)]# here dont use and , use & insted 
df1

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
23      1/1/2012 23:00     5.3               2.0         79               30   
24       1/2/2012 0:00     5.2               1.5         77               35   
25       1/2/2012 1:00     4.6               0.0         72               39   
26       1/2/2012 2:00     3.9              -0.9         71               32   
27       1/2/2012 3:00     3.7              -1.5         69               33   
...                ...     ...               ...        ...              ...   
8754  12/30/2012 18:00   -12.6             -16.0         76               24   
8755  12/30/2012 19:00   -13.4             -16.5         77               26   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   
8759  12/30/2012 23:00   -12.1             -15.1         78               28   
8760   12/31/2012 0:00   -11.1             -14.4         77               26   

      Visibility_km  Press_kPa Weather Condition  
23             25.0      99.31            Cloudy  
24             25.0      99.26      Rain Showers  
25             25.0      99.26            Cloudy  
26             25.0      99.26     Mostly Cloudy  
27             25.0      99.30     Mostly Cloudy  
...             ...        ...               ...  
8754           25.0     101.36      Mainly Clear  
8755           25.0     101.47      Mainly Clear  
8756           25.0     101.52             Clear  
8759           25.0     101.52     Mostly Cloudy  
8760           25.0     101.51            Cloudy  

[415 rows x 8 columns]

In [29]:
# Q.11 what is the mean value of each column against each 'weather condition'

In [32]:
data.groupby("Weather Condition").mean()

Temp_C  Dew Point Temp_C  \
Weather Condition                                                      
Clear                                     6.825716          0.089367   
Cloudy                                    7.970544          2.375810   
Drizzle                                   7.353659          5.504878   
Drizzle,Fog                               8.067500          7.033750   
Drizzle,Ice Pellets,Fog                   0.400000         -0.700000   
Drizzle,Snow                              1.050000          0.150000   
Drizzle,Snow,Fog                          0.693333          0.120000   
Fog                                       4.303333          3.159333   
Freezing Drizzle                         -5.657143         -8.000000   
Freezing Drizzle,Fog                     -2.533333         -4.183333   
Freezing Drizzle,Haze                    -5.433333         -8.000000   
Freezing Drizzle,Snow                    -5.109091         -7.072727   
Freezing Fog                             -7.575000         -9.250000   
Freezing Rain                            -3.885714         -6.078571   
Freezing Rain,Fog                        -2.225000         -3.750000   
Freezing Rain,Haze                       -4.900000         -7.450000   
Freezing Rain,Ice Pellets,Fog            -2.600000         -3.700000   
Freezing Rain,Snow Grains                -5.000000         -7.300000   
Haze                                     -0.200000         -2.975000   
Mainly Clear                             12.558927          4.581671   
Moderate Rain,Fog                         1.700000          0.800000   
Moderate Snow                            -5.525000         -7.250000   
Moderate Snow,Blowing Snow               -5.450000         -6.500000   
Mostly Cloudy                            10.574287          3.131174   
Rain                                      9.786275          7.042810   
Rain Showers                             13.722340          9.187766   
Rain Showers,Fog                         12.800000         12.100000   
Rain Showers,Snow Showers                 2.150000         -1.500000   
Rain,Fog                                  8.273276          7.219828   
Rain,Haze                                 4.633333          2.066667   
Rain,Ice Pellets                          0.600000         -0.600000   
Rain,Snow                                 1.055556         -0.566667   
Rain,Snow Grains                          1.900000         -2.100000   
Rain,Snow,Fog                             0.800000          0.300000   
Rain,Snow,Ice Pellets                     1.100000         -0.175000   
Snow                                     -4.524103         -7.623333   
Snow Pellets                              0.700000         -6.400000   
Snow Showers                             -3.506667         -7.866667   
Snow Showers,Fog                        -10.675000        -11.900000   
Snow,Blowing Snow                        -5.410526         -7.621053   
Snow,Fog                                 -5.075676         -6.364865   
Snow,Haze                                -4.020000         -6.860000   
Snow,Ice Pellets                         -1.883333         -3.666667   
Thunderstorms                            24.150000         19.750000   
Thunderstorms,Heavy Rain Showers         10.900000          9.000000   
Thunderstorms,Moderate Rain Showers,Fog  19.600000         18.500000   
Thunderstorms,Rain                       20.433333         18.533333   
Thunderstorms,Rain Showers               20.037500         17.618750   
Thunderstorms,Rain Showers,Fog           21.600000         18.700000   
Thunderstorms,Rain,Fog                   20.600000         18.600000   

                                         Rel Hum_%  Wind Speed_km/h  \
Weather Condition                                                     
Clear                                    64.497738        10.557315   
Cloudy                                   69.592593        16.127315   
Drizzle        

# Q.12  What is the minimum and maximum value of each column  aginst each weather condition?

In [34]:
data.groupby('Weather Condition').max()

Temp_C  Dew Point Temp_C  Rel Hum_%  \
Weather Condition                                                              
Clear                                      32.8              20.4         99   
Cloudy                                     30.5              22.6         99   
Drizzle                                    18.8              17.7         96   
Drizzle,Fog                                19.9              19.1        100   
Drizzle,Ice Pellets,Fog                     0.4              -0.7         92   
Drizzle,Snow                                1.2               0.2         95   
Drizzle,Snow,Fog                            1.1               0.6         98   
Fog                                        20.8              19.6        100   
Freezing Drizzle                           -2.3              -3.3         93   
Freezing Drizzle,Fog                       -0.3              -2.3         94   
Freezing Drizzle,Haze                      -5.0              -7.7         83   
Freezing Drizzle,Snow                      -3.3              -4.6         94   
Freezing Fog                               -0.1              -0.3         99   
Freezing Rain                               0.3              -1.7         92   
Freezing Rain,Fog                           0.1              -0.9         93   
Freezing Rain,Haze                         -4.9              -7.4         83   
Freezing Rain,Ice Pellets,Fog              -2.6              -3.7         92   
Freezing Rain,Snow Grains                  -5.0              -7.3         84   
Haze                                       14.1              11.1         86   
Mainly Clear                               33.0              21.2         99   
Moderate Rain,Fog                           1.7               0.8         94   
Moderate Snow                              -4.9              -6.7         93   
Moderate Snow,Blowing Snow                 -5.4              -6.4         93   
Mostly Cloudy                              32.4              24.4        100   
Rain                                       22.8              20.4         99   
Rain Showers                               26.4              23.0         97   
Rain Showers,Fog                           12.8              12.1         96   
Rain Showers,Snow Showers                   2.2              -1.2         78   
Rain,Fog                                   21.7              19.5        100   
Rain,Haze                                   5.5               2.9         86   
Rain,Ice Pellets                            0.6              -0.6         92   
Rain,Snow                                   1.7               0.5         94   
Rain,Snow Grains                            1.9              -2.1         75   
Rain,Snow,Fog                               0.8               0.3         96   
Rain,Snow,Ice Pellets                       1.3               0.1         94   
Snow                                        3.7               0.3         96   
Snow Pellets                                0.7              -6.4         59   
Snow Showers                                2.9              -0.7         94   
Snow Showers,Fog                          -10.0             -11.1         92   
Snow,Blowing Snow                          -1.4              -2.9         91   
Snow,Fog                                    1.1               0.8         99   
Snow,Haze                                  -3.6              -6.4         81   
Snow,Ice Pellets                            0.8              -1.7         92   
Thunderstorms                              26.7              20.1         87   
Thunderstorms,Heavy Rain Showers           10.9               9.0         88   
Thunderstorms,Moderate Rain Showers,Fog    19.6              18.5         93   
Thunderstorms,Rain                         21.3              19.1         93   
Thunderstorms,Rain Showers                 25.5              23.1         98   
Thunderstorms,Rain Showers,Fog           

In [35]:
data.groupby('Weather Condition').min()

Temp_C  Dew Point Temp_C  Rel Hum_%  \
Weather Condition                                                              
Clear                                     -23.3             -28.5         20   
Cloudy                                    -21.4             -26.8         18   
Drizzle                                     1.1              -0.2         74   
Drizzle,Fog                                 0.0              -1.6         85   
Drizzle,Ice Pellets,Fog                     0.4              -0.7         92   
Drizzle,Snow                                0.9               0.1         92   
Drizzle,Snow,Fog                            0.3              -0.1         92   
Fog                                       -16.0             -17.2         80   
Freezing Drizzle                           -9.0             -12.2         78   
Freezing Drizzle,Fog                       -6.4              -9.0         82   
Freezing Drizzle,Haze                      -5.8              -8.3         81   
Freezing Drizzle,Snow                      -8.3             -10.4         79   
Freezing Fog                              -19.0             -22.9         71   
Freezing Rain                              -6.5              -9.0         81   
Freezing Rain,Fog                          -6.1              -8.7         82   
Freezing Rain,Haze                         -4.9              -7.5         82   
Freezing Rain,Ice Pellets,Fog              -2.6              -3.7         92   
Freezing Rain,Snow Grains                  -5.0              -7.3         84   
Haze                                      -11.5             -16.0         68   
Mainly Clear                              -22.8             -28.0         20   
Moderate Rain,Fog                           1.7               0.8         94   
Moderate Snow                              -6.3              -7.6         83   
Moderate Snow,Blowing Snow                 -5.5              -6.6         92   
Mostly Cloudy                             -23.2             -28.5         18   
Rain                                        0.3              -5.7         40   
Rain Showers                                1.6              -7.2         37   
Rain Showers,Fog                           12.8              12.1         96   
Rain Showers,Snow Showers                   2.1              -1.8         75   
Rain,Fog                                    0.0              -1.2         83   
Rain,Haze                                   4.0               1.0         81   
Rain,Ice Pellets                            0.6              -0.6         92   
Rain,Snow                                   0.6              -1.7         81   
Rain,Snow Grains                            1.9              -2.1         75   
Rain,Snow,Fog                               0.8               0.3         96   
Rain,Snow,Ice Pellets                       0.9              -0.7         88   
Snow                                      -16.7             -24.6         41   
Snow Pellets                                0.7              -6.4         59   
Snow Showers                              -13.3             -19.3         52   
Snow Showers,Fog                          -11.3             -12.7         89   
Snow,Blowing Snow                         -12.0             -16.2         70   
Snow,Fog                                  -10.1             -12.0         77   
Snow,Haze                                  -4.3              -7.2         80   
Snow,Ice Pellets                           -4.3              -5.9         76   
Thunderstorms                              21.6              19.4         67   
Thunderstorms,Heavy Rain Showers           10.9               9.0         88   
Thunderstorms,Moderate Rain Showers,Fog    19.6              18.5         93   
Thunderstorms,Rain                         19.4              18.2         83   
Thunderstorms,Rain Showers                 11.0               7.0         68   
Thunderstorms,Rain Showers,Fog           

In [38]:
# Q.13 Show all the records where  weather  condition is fog
data1= data[data['Weather Condition']=='Fog']
data1

Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  Visibility_km  \
0       -1.8              -3.9         86                4            8.0   
1       -1.8              -3.7         87                4            8.0   
4       -1.5              -3.3         88                7            4.8   
5       -1.4              -3.3         87                9            6.4   
6       -1.5              -3.1         89                7            6.4   
...      ...               ...        ...              ...            ...   
8716   -16.0             -17.2         90                6            9.7   
8717   -14.8             -15.9         91                4            6.4   
8718   -13.8             -15.3         88                4            9.7   
8719   -14.8             -16.4         88                7            8.0   
8722   -12.0             -13.3         90                7            6.4   

      Press_kPa Weather Condition  
0        101.24               Fog  
1        101.24               Fog  
4        101.23               Fog  
5        101.27               Fog  
6        101.29               Fog  
...         ...               ...  
8716     101.25               Fog  
8717     101.25               Fog  
8718     101.25               Fog  
8719     101.22               Fog  
8722     101.15               Fog  

[150 rows x 7 columns]

In [53]:
#Q.14 Find all the instances where 'weather is clear' or 'visiblity' is  above 40
data1= data[(data['Weather Condition']=='Clear')| (data['Visibility_km']>40)]# | used here insted of or

data1

Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  Visibility_km  \
67     -16.9             -24.8         50               24           25.0   
106     -6.0             -10.0         73               17           48.3   
107     -5.6             -10.2         70               22           48.3   
108     -4.7              -9.6         69               20           48.3   
109     -4.4              -9.7         66               26           48.3   
...      ...               ...        ...              ...            ...   
8749   -12.4             -16.2         73               37           48.3   
8750   -11.8             -16.1         70               37           48.3   
8751   -11.3             -15.6         70               32           48.3   
8752   -11.4             -15.5         72               26           48.3   
8756   -13.8             -16.5         80               24           25.0   

      Press_kPa Weather Condition  
67       101.74             Clear  
106      100.45      Mainly Clear  
107      100.41      Mainly Clear  
108      100.38      Mainly Clear  
109      100.40      Mainly Clear  
...         ...               ...  
8749     100.92     Mostly Cloudy  
8750     100.96      Mainly Clear  
8751     101.05      Mainly Clear  
8752     101.15      Mainly Clear  
8756     101.52             Clear  

[3027 rows x 7 columns]

In [ ]:
#Q 15. Find all instances when:
#A. weather is clear and relative humidity is grater than 50 or 
#B visiblity is above 40


In [55]:
data1=data[(data["Weather Condition"]=='Clear')& (data['Rel Hum_%']>50)| (data['Visibility_km']>40)]
data1

Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  Visibility_km  \
106     -6.0             -10.0         73               17           48.3   
107     -5.6             -10.2         70               22           48.3   
108     -4.7              -9.6         69               20           48.3   
109     -4.4              -9.7         66               26           48.3   
110     -5.1             -10.7         65               22           48.3   
...      ...               ...        ...              ...            ...   
8749   -12.4             -16.2         73               37           48.3   
8750   -11.8             -16.1         70               37           48.3   
8751   -11.3             -15.6         70               32           48.3   
8752   -11.4             -15.5         72               26           48.3   
8756   -13.8             -16.5         80               24           25.0   

      Press_kPa Weather Condition  
106      100.45      Mainly Clear  
107      100.41      Mainly Clear  
108      100.38      Mainly Clear  
109      100.40      Mainly Clear  
110      100.46      Mainly Clear  
...         ...               ...  
8749     100.92     Mostly Cloudy  
8750     100.96      Mainly Clear  
8751     101.05      Mainly Clear  
8752     101.15      Mainly Clear  
8756     101.52             Clear  

[2921 rows x 7 columns]